### PR TITLE
diff 

### PR DESCRIPTION
**Pull request by anurag629**

### CODE REVIEW

**Comment by anurag629** [comment]
hey, @simonetgordon how to fix failing 'test-array-api'.   Thanks!

**Comment by anurag629** [comment]
> hey, @simonetgordon how to fix failing 'test-array-api'. Thanks!

**Review by simonetgordon — CHANGES_REQUESTED** [review]
Hi @anurag629, thanks for the contribution!

Could you link the issue you are addressing in this PR by commenting 'Closes #<issue_number>'? This helps us keep track of everything that has been addressed.

As for the failing array-api tests, there has been recent changes to these so you may want to pull from master to see if that fixes any. Also, you have some merge conflicts with master to resolve.

Thanks!

Simone

**Comment by anurag629** [comment]
Closes #6559

**Comment by anurag629** [comment]
> Hi @anurag629, thanks for the contribution!
> 
> Could you link the issue you are addressing in this PR by commenting 'Closes #<issue_number>'? This helps us keep track of everything that has been addressed.
> 
> As for the failing array-api tests, there has been recent changes to these so you may want to pull from master to see if that fixes any. Also, you have some merge conflicts with master to resolve.
> 
> Thanks!
> 
> Simone

Hi @simonetgordon, I have resolved conflicts and addressed this PR to the respective issue. Please review it again.

Thanks!

**Review by anurag629 — COMMENTED** [review]
self review

**Comment by anurag629** [comment]
@simonetgordon  why these 'test-array-api (....., linalg)' test failing?
Thanks!

File: ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py [hunk 1]
```diff
@@ -2482,3 +2482,44 @@ def test_isreal(
...
+        max_num_dims=3,
+        min_value=-100,
+        max_value=100,
+        allow_nan=False,
+    ),
+    num_positional_args=helpers.num_positional_args(fn_name="diff"),
+)
+def test_diff(
+    dtype_and_x,
+    with_out,
```
Comment (simonetgordon) [inline ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py:2504]: Need to define strategies for `with_out`, `as_variable` etc in the `@given` decorator

**Review by simonetgordon — CHANGES_REQUESTED** [review]

**Comment by simonetgordon** [comment]
> > Hi @anurag629, thanks for the contribution!
> > Could you link the issue you are addressing in this PR by commenting 'Closes #<issue_number>'? This helps us keep track of everything that has been addressed.
> > As for the failing array-api tests, there has been recent changes to these so you may want to pull from master to see if that fixes any. Also, you have some merge conflicts with master to resolve.
> > Thanks!
> > Simone
> 
> Hi @simonetgordon, I have resolved conflicts and addressed this PR to the respective issue. Please review it again. Should I also remove unused imports which are causing lint checks? Thanks!

Yes please do remove the imports and address any other issues in the files you have altered which are causing the lint check to fail.

The array API tests do not need to run on this PR, you can ignore them. I have replaced the array-API label with the experimental label so the `test_diff` should run now.

File: ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py (hunk 1 above)
Comment (anurag629) [reply ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py:2505]: Hi, @simonetgordon added strategies in <code>@given</code> decoder. Please check if it's correct.
Thanks!

**Comment by anurag629** [comment]
@simonetgordon Please review this PR.
Thanks!

**Review by simonetgordon — CHANGES_REQUESTED** [review]
diff appears twice in the jax elementwise backend, it should only be in the `experimental/` directory.

**Comment by simonetgordon** [comment]
> @simonetgordon Please review this PR. Thanks!

Hi there, please inspect the error output for `test_diff` to try and understand what you need to fix:

```
=================================== FAILURES ===================================
__________ test_diff[cpu-ivy.functional.backends.jax-False-False-jax] __________
ivy_tests/test_ivy/helpers/testing_helpers.py:184: in new_fn
    @settings(max_examples=1)
ivy_tests/test_ivy/helpers/testing_helpers.py:226: in new_fn
    return test_fn(*args, **kwargs)
ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py:2490: in test_diff
    @given(
ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py:2514: in test_diff
    helpers.test_function(
ivy_tests/test_ivy/helpers/function_testing.py:281: in test_function
    ret, ret_np_flat = get_ret_and_flattened_np_array(
ivy_tests/test_ivy/helpers/function_testing.py:1617: in get_ret_and_flattened_np_array
    ret = fn(*args, **kwargs)
ivy/func_wrapper.py:415: in new_fn
    return fn(*args, **kwargs)
ivy/func_wrapper.py:360: in new_fn
    return fn(*args, **kwargs)
ivy/func_wrapper.py:84: in new_fn
    return fn(*new_args, **new_kwargs)
ivy/func_wrapper.py:146: in new_fn
    ret = fn(*args, **kwargs)
ivy/functional/backends/jax/experimental/elementwise.py:186: in diff
    return jnp.diff(x1, x2)
E   ValueError: Non-hashable static arguments are not supported. An error occurred during a call to 'diff' while trying to hash an object of type <class 'jaxlib.xla_extension.DeviceArray'>, [-1]. The error was:
E   TypeError: unhashable type: 'DeviceArray'
E   
E   Falsifying example: test_diff(
E       dtype_and_x=(['bfloat16', 'bfloat16'],
E        [array([-1], dtype=bfloat16), array([-1], dtype=bfloat16)]),
E       num_positional_args=2,
E       with_out=False,
E       as_variable=[False],
E       native_array=[False],
E       container=[False],
E       instance_method=False,
E       fw='jax',
E   )
E   
E   You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AXicY2BkAAMoBaaR2AAAcQAG') as a decorator on your test case
E   Falsifying example: new_fn(
E       data=data(...),
E       get_command_line_flags={'as_variable': None,
E        'native_array': None,
E        'with_out': None,
E        'container': None,
E        'instance_method': None,
E        'test_gradients': None},
E       device='cpu',
E       f=<module 'ivy.functional.backends.jax' from '/ivy/ivy/functional/backends/jax/__init__.py'>,
E       fw='jax',
E       fixt_frontend_str=None,
E   )
E   Draw 1: False
E   Draw 2: False
E   Draw 3: False
E   Draw 4: False
E   Draw 5: False
E   
E   You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AAAAAAAA') as a decorator on your test case
=============================== warnings summary ===============================
../usr/local/lib/python3.8/dist-packages/flatbuffers/compat.py:19
  /usr/local/lib/python3.8/dist-packages/flatbuffers/compat.py:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

ivy/container/base.py:12[66](https://github.com/unifyai/ivy/actions/runs/3442651086/jobs/5743441259#step:7:67)
  /ivy/ivy/container/base.py:1266: DeprecationWarning: invalid escape sequence \.
    flat_keys = re.split("/|\.", key_chain)  # noqa

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py::test_diff[cpu-ivy.functional.backends.jax-False-False-jax]
````

**Comment by anurag629** [comment]
> > @simonetgordon Please review this PR. Thanks!
> 
> Hi there, please inspect the error output for `test_diff` to try and understand what you need to fix:
> 
> ```
> =================================== FAILURES ===================================
> __________ test_diff[cpu-ivy.functional.backends.jax-False-False-jax] __________
> ivy_tests/test_ivy/helpers/testing_helpers.py:184: in new_fn
>     @settings(max_examples=1)
> ivy_tests/test_ivy/helpers/testing_helpers.py:226: in new_fn
>     return test_fn(*args, **kwargs)
> ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py:2490: in test_diff
>     @given(
> ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py:2514: in test_diff
>     helpers.test_function(
> ivy_tests/test_ivy/helpers/function_testing.py:281: in test_function
>     ret, ret_np_flat = get_ret_and_flattened_np_array(
> ivy_tests/test_ivy/helpers/function_testing.py:1617: in get_ret_and_flattened_np_array
>     ret = fn(*args, **kwargs)
> ivy/func_wrapper.py:415: in new_fn
>     return fn(*args, **kwargs)
> ivy/func_wrapper.py:360: in new_fn
>     return fn(*args, **kwargs)
> ivy/func_wrapper.py:84: in new_fn
>     return fn(*new_args, **new_kwargs)
> ivy/func_wrapper.py:146: in new_fn
>     ret = fn(*args, **kwargs)
> ivy/functional/backends/jax/experimental/elementwise.py:186: in diff
>     return jnp.diff(x1, x2)
> E   ValueError: Non-hashable static arguments are not supported. An error occurred during a call to 'diff' while trying to hash an object of type <class 'jaxlib.xla_extension.DeviceArray'>, [-1]. The error was:
> E   TypeError: unhashable type: 'DeviceArray'
> E   
> E   Falsifying example: test_diff(
> E       dtype_and_x=(['bfloat16', 'bfloat16'],
> E        [array([-1], dtype=bfloat16), array([-1], dtype=bfloat16)]),
> E       num_positional_args=2,
> E       with_out=False,
> E       as_variable=[False],
> E       native_array=[False],
> E       container=[False],
> E       instance_method=False,
> E       fw='jax',
> E   )
> E   
> E   You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AXicY2BkAAMoBaaR2AAAcQAG') as a decorator on your test case
> E   Falsifying example: new_fn(
> E       data=data(...),
> E       get_command_line_flags={'as_variable': None,
> E        'native_array': None,
> E        'with_out': None,
> E        'container': None,
> E        'instance_method': None,
> E        'test_gradients': None},
> E       device='cpu',
> E       f=<module 'ivy.functional.backends.jax' from '/ivy/ivy/functional/backends/jax/__init__.py'>,
> E       fw='jax',
> E       fixt_frontend_str=None,
> E   )
> E   Draw 1: False
> E   Draw 2: False
> E   Draw 3: False
> E   Draw 4: False
> E   Draw 5: False
> E   
> E   You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AAAAAAAA') as a decorator on your test case
> =============================== warnings summary ===============================
> ../usr/local/lib/python3.8/dist-packages/flatbuffers/compat.py:19
>   /usr/local/lib/python3.8/dist-packages/flatbuffers/compat.py:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
>     import imp
> 
> ivy/container/base.py:12[66](https://github.com/unifyai/ivy/actions/runs/3442651086/jobs/5743441259#step:7:67)
>   /ivy/ivy/container/base.py:1266: DeprecationWarning: invalid escape sequence \.
>     flat_keys = re.split("/|\.", key_chain)  # noqa
> 
> -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
> =========================== short test summary info ============================
> FAILED ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py::test_diff[cpu-ivy.functional.backends.jax-False-False-jax]
> ```

Hey @simonetgordon, I corrected the issue. This error is bcoz I was passing two array but 'diff' only accepts 1 array. Now, I fixed that. 

Please review it again! :)

File: ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py [hunk 2]
```diff
@@ -2483,3 +2483,39 @@ def test_isreal(
...
+# diff
+@handle_cmd_line_args
+@given(
+    dtype_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("numeric"),
+        num_arrays=1,
+    ),
+    num_positional_args=helpers.num_positional_args(fn_name="diff"),
+)
+def test_diff(
```
Comment (simonetgordon) [inline ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py:2497]: `test_diff` should be in `test_functional/test_experimental/test_core/test_elementwise.py`

**Review by simonetgordon — CHANGES_REQUESTED** [review]

**Comment by simonetgordon** [comment]
Hi there, can you explain to me what the error is and your confusion with it?

File: ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py (hunk 2 above)
Comment (anurag629) [reply ivy_tests/test_ivy/test_functional/test_core/test_elementwise.py]: moved test_diff to the correct place

**Comment by anurag629** [comment]
> Hi there, can you explain to me what the error is and your confusion with it?

Hey, got the error. I was passing an extra argument to numpy.diff(). Now the test passed. 
Thanks!

**Review by simonetgordon — COMMENTED** [review]
Hi @anurag629, provisionally everything looks good, but unfortunately I cannot find `test_diff` running on the PR checks. Does it run locally for you OK?

I will look into it and try and get it fixed.

**Comment by anurag629** [comment]
> Hi @anurag629, provisionally everything looks good, but unfortunately I cannot find `test_diff` running on the PR checks. Does it run locally for you OK?
> 
> I will look into it and try and get it fixed.

Yes, its running locally fine.

**Comment by anurag629** [comment]
> Hi @anurag629, provisionally everything looks good, but unfortunately I cannot find `test_diff` running on the PR checks. Does it run locally for you OK?
> 
> I will look into it and try and get it fixed.

Yesterday,  I checked the test are showing but now they are not.
Screenshot from 2022-11-15 11-40-15

**Comment by anurag629** [comment]
Hey @simonetgordon Please review this PR

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py [hunk 3]
```diff
@@ -884,3 +884,40 @@ def test_nextafter(
...
+    dtype_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
+        min_num_dims=1,
+        max_num_dims=3,
+        min_dim_size=1,
+        max_dim_size=3,
+    ),
+    num_positional_args=helpers.num_positional_args(fn_name="diff"),
+)
+def test_diff(
```
Comment (simonetgordon) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py:901]: The check is now running for `test_diff` and this is the error message:
```
/ivy/ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py:850
__ ERROR at setup of test_diff[cpu-ivy.functional.backends.torch-False-False] __
file /ivy/ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py, line 890
  @handle_test(
      fn_tree="functional.experimental.diff",
      dtype_and_x=helpers.dtype_and_values(
          available_dtypes=helpers.get_dtypes("float"),
          min_num_dims=1,
          max_num_dims=3,
          min_dim_size=1,
          max_dim_size=3,
      ),
      num_positional_args=helpers.num_positional_args(fn_name="diff"),
  )
  def test_diff(
E       fixture 'container' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, include_metadata_in_junit_xml, json_metadata, metadata, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, run_around_tests, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/ivy/ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py:890
```

Comment (simonetgordon) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py:901]: Perhaps take a look at the other `experimental` test functions that passed and see if you can find some differences

**Review by simonetgordon — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py (hunk 3 above)
Comment (anurag629) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py:903]: Hey, @simonetgordon how to resolve this error? I tried but was not able to solve it. I also checked for other available tests but got the same error.

Screenshot from 2022-11-20 14-01-31

Comment (simonetgordon) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py:904]: @anurag629 Hi there. I actually encountered this issue myself earlier in the week. I believe it is related to accidental changes to `.idea` folder configuration files being pushed to master recently. I tried fixing a couple of the issues in the configuration files that I noticed (e.g. typos) but it didn't seem to work. In the end, I deleted my local Ivy repo and set up again i.e. re-cloned etc. Then there were only the following files in the `.idea` directory:
```
inspectionProfiles
ivy.iml
misc.xml
modules.xml
vcs.xml
workspace.xml
```

You could either do as I did and delete your local code base and re-clone, or you could try deleting the configuration files which don't appear on the list above.

Comment (simonetgordon) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py:904]: Apologies for the issues by the way, let me know how you progress with fixing it!

In the meantime, you can still inspect the output of the test function via the details of the PR check i.e. `test-experimental-core-ivy` -> `numpy, elementwise` -> 'Run tests' and search the error output for `test_diff`

Comment (anurag629) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_elementwise.py:906]: Hi @simonetgordon Now tests are failing due to the zeta function.

https://github.com/unifyai/ivy/actions/runs/3510006434/jobs/5879494232#step:7:59

Screenshot from 2022-11-21 03-55-11

**Comment by anurag629** [comment]
@simonetgordon please review this PR.

**Review by simonetgordon — APPROVED** [review]
`test_diff` passing, ready to be merged! Thanks!

**Pull request merged**